src/sagas/tokens.js

```diff
@@ -0,0 +1,291 @@
+
+  try {
+    // Retry mechanism
+    for (let i = 0; i <= METADATA_MAX_RETRIES; i += 1) {
+      try {
+        const data = yield call(metadataApi.getDagMetadata, tokenId, network);
+
+        yield put({
+          type: types.TOKEN_FETCH_METADATA_SUCCESS,
+          tokenId,
+          data: get(data, tokenId, null),
+        });
+        return;
+      } catch (e) {
+        yield delay(1000); // Wait 1s before trying again
+      }
+    }
+
+    throw new Error(`Max retries requesting metadata for ${tokenId}`);
+  } catch (e) {
+    yield put({
+      type: types.TOKEN_FETCH_METADATA_FAILED,
+      tokenId,
+    });
+    // eslint-disable-next-line
+    console.log('Error downloading metadata of token', tokenId);
+  }
+}
+
+function* fetchTokenHistory(action) {
+  const { tokenId, force } = action;
+
+  try {
+    const wallet = yield select((state) => state.wallet);
+    const tokenHistory = yield select((state) => get(state.tokensHistory, tokenId));
+
+    if (!force && tokenHistory && tokenHistory.oldStatus === TOKEN_DOWNLOAD_STATUS.READY) {
+      // The data is already loaded, we should dispatch success
+      yield put(tokenFetchHistorySuccess(tokenId, tokenHistory.data));
+      return;
+    }
+
+    const response = yield call(wallet.getTxHistory.bind(wallet), { token_id: tokenId });
+    const data = response.map((txHistory) => helpers.mapTokenHistory(txHistory, tokenId));
+
+    yield put(tokenFetchHistorySuccess(tokenId, data));
+  } catch (e) {
+    yield put(tokenFetchHistoryFailed(tokenId));
+  }
+}
+
+/**
+ * This saga will route the actions dispatched from SET_TOKEN and NEW_TOKEN to the
+ * TOKEN_FETCH_BALANCE_REQUESTED saga, the idea is to load the balance for new tokens
+ * registered or created on the app.
+ */
+function* routeTokenChange(action) {
+  const wallet = yield select((state) => state.wallet);
+
+  if (!wallet || !wallet.isReady()) {
+    return;
+  }
+
+  switch (action.type) {
+    default:
+    case 'new_tokens':
+      for (const token of action.payload.tokens) {
+        yield put({
+          type: types.TOKEN_FETCH_BALANCE_REQUESTED,
+          tokenId: token.uid,
+        });
+      }
+      break;
+  }
+}
+
+export function* fetchTokenData(tokenId) {
+  const fetchBalanceResponse = yield call(
+    dispatchAndWait,
+    tokenFetchBalanceRequested(tokenId, true),
+    specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_SUCCESS, {
+      tokenId,
+    }),
+    specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_FAILED, {
+      tokenId,
+    }),
+  );
+  const fetchHistoryResponse = yield call(
+    dispatchAndWait,
+    tokenFetchHistoryRequested(tokenId, true),
+    specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_SUCCESS, {
+      tokenId,
+    }),
+    specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_FAILED, {
+      tokenId,
+    }),
+  );
+
+  if (fetchHistoryResponse.failure || fetchBalanceResponse.failure) {
+    throw new Error(`Error loading HTR history or balance for token ${tokenId}`);
```

Comment:
Why HTR?

Reply:
Not sure, removed in 5443b92d63cc0e78fbf203ded51f2281918d1f38